File: Algorithms/Sorting/Quicksort/Quicksort.cpp
template <class T>
void quicksort(T arr[], int low, int high) {
  if (low >= high)
    return;

  int left = low, right = high, pivot = arr[(low + high) / 2];
  while (left <= right) {
    while (arr[left] < pivot)
      left++;
    while (arr[right] > pivot)
      right--;

    if (left <= right) {
      int temp = arr[left];
      arr[left++] = arr[right];
      arr[right--] = temp;
    }
  }

  quicksort<T>(arr, low, right);
  quicksort<T>(arr, left, high);
}

File: Algorithms/Sorting/Quicksort/Quicksort.py
def quicksort(arr, low, high):
    if low >= high:
        return

    left, right, pivot = low, high, arr[(low + high) // 2]
    while left <= right:
        while arr[left] < pivot:
            left += 1
        while arr[right] > pivot:
            right -= 1

        if left <= right:
            arr[left], arr[right] = arr[right], arr[left]
            left += 1
            right -= 1

    quicksort(arr, low, right)
    quicksort(arr, left, high)
